integrations/spark/src/main/java/marquez/spark/agent/lifecycle/plan/UnknownEntryFacetListener.java

```diff
@@ -0,0 +1,60 @@
+package marquez.spark.agent.lifecycle.plan;
+
+import static java.util.Optional.ofNullable;
+import static scala.collection.JavaConversions.mapAsJavaMap;
+import static scala.collection.JavaConversions.seqAsJavaList;
+
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import marquez.spark.agent.facets.UnknownEntryFacet;
+import marquez.spark.agent.facets.UnknownEntryFacet.FacetEntry;
+import org.apache.spark.sql.catalyst.expressions.AttributeReference;
+import org.apache.spark.sql.catalyst.expressions.AttributeSet;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import org.apache.spark.sql.types.DataType;
+import scala.collection.JavaConversions;
+
+public class UnknownEntryFacetListener implements Consumer<LogicalPlan> {
+
+  private final Map<LogicalPlan, Object> processedNodes = new IdentityHashMap<>();
+
+  @Override
+  public void accept(LogicalPlan logicalPlan) {
+    processedNodes.put(logicalPlan, null);
+  }
+
+  public UnknownEntryFacet build(LogicalPlan root) {
+    FacetEntry output = mapEntry(root);
+    List<FacetEntry> inputs =
+        seqAsJavaList(root.collectLeaves()).stream()
+            .map(this::mapEntry)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+    return output == null && inputs.isEmpty() ? null : new UnknownEntryFacet(output, inputs);
+  }
+
+  private FacetEntry mapEntry(LogicalPlan x) {
+    if (processedNodes.containsKey(x)) return null;
```

Comment:
I'm not sure I understand when `mapEntry` will ever get past the first line. If I read the `PlanTraversal` code correctly, the `accpet` method here gets called for every visited node when we call `collect` on the LogicalPlan. In what circumstances will we not visit a node that gets returned by the `collectLeaves` method call here in `build`?

Reply:
we will not visit a node if we don't have partial function which is going to map it.
At every `PartialFunction` we have method `isDefinedAt`, which is saying that this function can map a Node to Dataset.
We should skip such Nodes, cause it's known for us.
In `accept` method we collect all "known" nodes, and then in `mapEntry` we have a check/filter to skip already visited nodes.
```
private FacetEntry mapEntry(LogicalPlan x) {
    if (processedNodes.containsKey(x)) return null;
```